File: stack/Generate Parentheses.py
def generateParenthesis(n: int) -> list[str]:
    """
    We have two options:
    Option 1: Add opening paren and n -= 1
    Option 2: Add closing paren iff stack has an opening paren
    Do this while backtracking both choices
    """

    ans = []

    def backtrack(stack, n, parens):
        if n == 0 and not stack:
            ans.append(parens)
            return

        if n > 0:
            stack.append('(')
            backtrack(stack, n - 1, parens + '(')
            stack.pop()
        if stack:
            stack.pop()
            backtrack(stack, n, parens + ')')
            stack.append('(')

    backtrack([], n, '')
    return ans

print(generateParenthesis(10))

File: stack/Generate Parentheses.cs
public class SolutionGenerateParenthesis
{
    private List<string> ans;
    private int parens;

    private void backtrack(int open, int close, string curr)
    {
        if (open > parens || close > parens)
            return;

        if (open == parens && close == parens)
        {
            ans.Add(curr);
            return;
        }

        if (close < open)
        {
            backtrack(open, close + 1, curr + ")");
            if (open < parens)
                backtrack(open + 1, close, curr + "(");

        }
        else
            backtrack(open + 1, close, curr + "(");
    }

    public IList<string> GenerateParenthesis(int n)
    {
        ans = new List<string>();
        parens = n;

        backtrack(0, 0, "");
        return ans;
    }
}
